File: Algorithm/Dutch National Flag/leetcode/75. Sort Colors/GPT/Sort-Colors.js
// # 1. 問題の分析

// ### 競技プログラミング視点

// * **速度最優先**

//   * sort禁止なので、最適解は **O(n), O(1) space** の1パス解法（Dutch National Flagアルゴリズム）。
//   * `low`, `mid`, `high` の3ポインタでインプレース交換しながら色を並べ替える。
// * **メモリ最小化**

//   * 補助配列を作らず、インデックス操作のみで解決。

// ### 業務開発視点

// * **可読性・保守性**

//   * `low`, `mid`, `high` の役割を明確化して命名。
//   * 入力検証を実装 → 想定外のデータに早期例外。
// * **例外方針**

//   * 型が配列でない → `TypeError`
//   * 要素が `0|1|2` 以外 → `RangeError`

// ### JavaScript特有の考慮点

// * **V8最適化**

//   * 単純な `for` / `while` ループで型を安定化。
//   * `Array.isArray` / `typeof` で軽量ガード。
// * **GC対策**

//   * 補助配列なし、クロージャなし。
// * **配列操作**

//   * `swap` をインライン実装。

// ---

// # 2. アルゴリズムアプローチ比較

// | アプローチ                            | 時間計算量 | 空間計算量 | JS実装コスト | 可読性 | 備考         |
// | -------------------------------- | ----- | ----- | ------- | --- | ---------- |
// | 方法A: Dutch National Flag (3ポインタ) | O(n)  | O(1)  | 低       | 中   | 最速・省メモリ    |
// | 方法B: カウントソート (2パス)               | O(n)  | O(1)  | 中       | 高   | 直感的だが2パス   |
// | 方法C: バブルソート/選択ソート                | O(n²) | O(1)  | 低       | 高   | 制約が小さい場合のみ |

// ---

// # 3. 選択したアルゴリズムと理由

// * **選択: 方法A（Dutch National Flag）**
// * **理由:**

//   * 制約 (n ≤ 300) ではどの方法でも十分だが、Follow-up要件「1パス・定数メモリ」に合致するのはこれ。
// * **JS最適化ポイント:**

//   * `while (mid <= high)` のループ
//   * `swap` をインラインで処理し一時オブジェクト生成なし
//   * `Number.isInteger` で入力チェック

// ---

// # 4. コード実装（solution.js）

// ```javascript
'use strict';

/**
 * Sort Colors (Dutch National Flag Problem)
 * @param {number[]} nums - 入力配列（0: red, 1: white, 2: blue）
 * @returns {void} in-place で nums を並び替え
 * @throws {TypeError} 入力が配列でない場合
 * @throws {RangeError} 要素が 0,1,2 以外の場合
 * @complexity 時間 O(n), 空間 O(1)
 */
var sortColors = function (nums) {
    // 入力検証
    if (!Array.isArray(nums)) {
        throw new TypeError('Input must be an array');
    }
    const n = nums.length;
    if (n < 1 || n > 300) {
        throw new RangeError('Array size out of allowed range (1 <= n <= 300)');
    }

    for (let i = 0; i < n; i++) {
        const v = nums[i];
        if (!Number.isInteger(v)) {
            throw new TypeError('Array must contain integers only');
        }
        if (v < 0 || v > 2) {
            throw new RangeError('Array values must be 0, 1, or 2');
        }
    }

    // Dutch National Flag アルゴリズム
    let low = 0;
    let mid = 0;
    let high = n - 1;

    while (mid <= high) {
        if (nums[mid] === 0) {
            // swap(nums[low], nums[mid])
            const tmp = nums[low];
            nums[low] = nums[mid];
            nums[mid] = tmp;
            low++;
            mid++;
        } else if (nums[mid] === 1) {
            mid++;
        } else {
            // nums[mid] === 2
            const tmp = nums[mid];
            nums[mid] = nums[high];
            nums[high] = tmp;
            high--;
        }
    }
};

module.exports = { sortColors };
// ```

// ---

// # 5. 追加メモ（JS最適化チェックリスト）

// * ✅ ループは `while` を使用、関数呼び出し削減。
// * ✅ swap はインライン、余分な関数オブジェクトを生成しない。
// * ✅ 型崩れを避けるため整数検証。
// * ✅ Pure（副作用は引数の破壊更新のみ、返り値なし → LeetCode準拠）。

// ---

// ✅ これで `leetcode` の提出フォーマットに完全対応（`var sortColors = function(nums) { ... }`）。
// ✅ Node.js v18 で `node solution.js` 実行可能（ただしLeetCode環境では `module.exports` は無視されます）。

// ---

File: Algorithm/Dutch National Flag/leetcode/75. Sort Colors/GPT/Sort-Colors.py
# ## 1. 問題分析

# ### 競技プログラミング視点

# * **実行速度**: 一度の走査 O(n) が最速。
# * **メモリ効率**: 定数領域 O(1)。
# * **Python制約**: インタープリターのループコストを抑えるため、組み込み swap 操作で実装。

# ### 業務開発視点

# * **可読性**: 「赤=0, 白=1, 青=2」を扱うことを明示。
# * **保守性**: 変数名を `low`, `mid`, `high` として直感的に。
# * **型ヒント**: `List[int]` を利用。
# * **エラーハンドリング**: 入力検証を追加。

# ### Python特有の考慮点

# * `list` アクセスは O(1) → インデックス操作で効率的。
# * ループ回数は最大 n → Python のオーバーヘッドは許容範囲。
# * 内包表記は不要（就地変更なので return しない）。

# ---

# ## 2. アルゴリズム比較

# | アプローチ                            | 時間計算量 | 空間計算量 | 実装コスト | 可読性 | Python標準ライブラリ活用 | 備考        |
# | -------------------------------- | ----- | ----- | ----- | --- | --------------- | --------- |
# | 方法A: Dutch National Flag (ポインタ法) | O(n)  | O(1)  | 中     | 高   | なし              | 最適解       |
# | 方法B: カウントソート                     | O(n)  | O(1)  | 中     | 中   | Counter         | 2回走査が必要   |
# | 方法C: バブル/選択ソート                   | O(n²) | O(1)  | 低     | 高   | なし              | 小規模専用、非推奨 |

# ---

# ## 3. 実装方針

# * **選択アプローチ**: 方法A（Dutch National Flag アルゴリズム）
# * **理由**:

#   * 計算量 O(n), 空間 O(1) → 最速・最省メモリ。
#   * Python でも swap 操作で簡潔に記述可能。
#   * 可読性を保ちながら競技・業務両対応。

# ---

# ## 4. コード実装例

# ```python
from typing import List


class Solution:
    def sortColors(self, nums: List[int]) -> None:
        """
        オランダ国旗問題の解法 (業務開発向け堅牢版)

        Args:
            nums (List[int]): 色を表す配列 (0=赤, 1=白, 2=青)
        Returns:
            None: 配列を in-place でソート
        Raises:
            ValueError: 入力が空、または 0/1/2 以外を含む場合
        """
        if not nums:
            raise ValueError("Input must not be empty")
        if any(num not in (0, 1, 2) for num in nums):
            raise ValueError("Input must contain only 0, 1, or 2")

        low, mid, high = 0, 0, len(nums) - 1
        while mid <= high:
            if nums[mid] == 0:
                nums[low], nums[mid] = nums[mid], nums[low]
                low += 1
                mid += 1
            elif nums[mid] == 1:
                mid += 1
            else:  # nums[mid] == 2
                nums[mid], nums[high] = nums[high], nums[mid]
                high -= 1

    def sortColors_optimized(self, nums: List[int]) -> None:
        """
        競技プログラミング向け最適化版
        - 入力検証を省略
        - 性能最優先
        """
        low, mid, high = 0, 0, len(nums) - 1
        while mid <= high:
            if nums[mid] == 0:
                nums[low], nums[mid] = nums[mid], nums[low]
                low += 1
                mid += 1
            elif nums[mid] == 1:
                mid += 1
            else:
                nums[mid], nums[high] = nums[high], nums[mid]
                high -= 1


# ```

# ---

# ## 5. 計算量まとめ

# * **時間計算量**: O(n)
# * **空間計算量**: O(1)
# * **Python固有オーバーヘッド**:

#   * ループの実行コストは O(n) で支配的。
#   * `swap` 操作はタプルアンパックで効率的に実行可能。

# ---

# ## 6. Python固有の最適化観点

# * **組み込み関数活用**: `any` を使って入力検証を簡潔化。
# * **不要コピー回避**: in-place で処理。
# * **型ヒント**: `List[int]` → Pylance による静的解析可能。
# * **slots やジェネレータ**: 今回は不要。

# ---

# ✅ 以上により、**業務向けには `sortColors` を使用、競プロでは `sortColors_optimized` を使用**する二段構成がベストです。
